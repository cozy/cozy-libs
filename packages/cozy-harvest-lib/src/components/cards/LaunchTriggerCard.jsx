import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import Button from 'cozy-ui/transpiled/react/Button'
import Card from 'cozy-ui/transpiled/react/Card'
import { translate } from 'cozy-ui/transpiled/react/I18n'
import Icon from 'cozy-ui/transpiled/react/Icon'
import { Text } from 'cozy-ui/transpiled/react/Text'

import TriggerLauncher from '../TriggerLauncher'

export class LaunchTriggerCard extends PureComponent {
  render() {
    const { className, trigger, t, ...rest } = this.props
    return (
      <Card className={className} {...rest}>
        <TriggerLauncher trigger={trigger}>
          {({ error, launch, running }) => (
            <div>
              {error && (
                <Text className="u-error">{t('card.launchTrigger.error')}</Text>
              )}
              <Button
                label={t('card.launchTrigger.button.label')}
                icon={<Icon focusable="false" icon="sync" spin={running} />}
                className="u-mv-half u-mh-0"
                disabled={running}
                onClick={launch}
                subtle
              />
            </div>
          )}
        </TriggerLauncher>
      </Card>
    )
  }
}

LaunchTriggerCard.propTypes = {
  ...Card.propTypes,
  trigger: PropTypes.object.isRequired,
  t: PropTypes.func.isRequired
}

export default translate()(LaunchTriggerCard)
