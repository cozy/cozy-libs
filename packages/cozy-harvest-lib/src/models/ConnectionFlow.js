import MicroEE from 'microee'

import Realtime from 'cozy-realtime'

import {
  fetchAccountsWithoutTriggers,
  saveAccount,
  ACCOUNTS_DOCTYPE
} from '../connections/accounts'
import clone from 'lodash/clone'
import {
  launchTrigger,
  prepareTriggerAccount,
  fetchTrigger,
  ensureTrigger
} from '../connections/triggers'
import { watchKonnectorJob } from '../models/konnector/KonnectorJobWatcher'
import logger from '../logger'
import { findKonnectorPolicy } from '../konnector-policies'
import { createOrUpdateCipher } from '../models/cipherUtils'
import assert from '../assert'
import * as accounts from '../helpers/accounts'
import * as triggersModel from '../helpers/triggers'

const JOBS_DOCTYPE = 'io.cozy.jobs'

// events
export const ERROR_EVENT = 'error'
export const LOGIN_SUCCESS_EVENT = 'loginSuccess'
export const SUCCESS_EVENT = 'success'
export const TRIGGER_LAUNCH_EVENT = 'TRIGGER_LAUNCH_EVENT'

export const TWO_FA_REQUEST_EVENT = 'twoFARequest'
export const TWO_FA_MISMATCH_EVENT = 'twoFAMismatch'
export const UPDATE_EVENT = 'update'

// statuses
export const IDLE = 'IDLE'
export const CREATING_ACCOUNT = 'CREATING_ACCOUNT'
export const PENDING = 'PENDING'
export const ERRORED = 'ERRORED'
export const LOGIN_SUCCESS = 'LOGIN_SUCCESS'
export const SUCCESS = 'SUCCESS'
export const RUNNING_TWOFA = 'RUNNING_TWOFA'
export const WAITING_TWOFA = 'WAITING_TWOFA'
export const TWO_FA_MISMATCH = 'TWO_FA_MISMATCH'
export const TWO_FA_REQUEST = 'TWO_FA_REQUEST'

const JOB_EVENTS = [ERROR_EVENT, LOGIN_SUCCESS_EVENT, SUCCESS_EVENT]

const eventToStatus = {
  [ERROR_EVENT]: ERRORED,
  [LOGIN_SUCCESS_EVENT]: LOGIN_SUCCESS,
  [SUCCESS_EVENT]: SUCCESS
}
const stepEvents = [LOGIN_SUCCESS_EVENT]
const isStepEvent = eventName => stepEvents.includes(eventName)

/**
 * Creates or updates an io.cozy.accounts from user submitted data
 * Used as a form submit handler
 *
 * @param  {io.cozy.account} options.account - Existing io.cozy.account or object
 * @param  {Cipher} options.cipher - Vault cipher if vault has been unlocked
 * @param  {CozyClient} options.client - A cozy client
 * @param  {io.cozy.konnector} options.konnector - Konnector to which the account is linked
 * @param  {KonnectorPolicy} options.konnectorPolicy - Controls if auth is saved in io.cozy.accounts
 * and if auth is saved into the vault
 * @param  {function} options.saveAccount
 * @param  {object} options.userCredentials
 */
export const createOrUpdateAccount = async ({
  account,
  cipher,
  client,
  flow,
  konnector,
  konnectorPolicy,
  userCredentials
}) => {
  assert(flow, 'No flow')
  const isUpdate = !!account

  if (isUpdate) {
    logger.debug('Updating the account...')
  } else {
    logger.debug('Creating the account...')
  }

  const { onAccountCreation, saveInVault } = konnectorPolicy

  let accountToSave = clone(account) || {}

  accountToSave = accounts.resetState(accountToSave)

  accountToSave = accounts.setSessionResetIfNecessary(
    accountToSave,
    userCredentials
  )

  if (isUpdate) {
    accountToSave = accounts.mergeAuth(accountToSave, userCredentials)
  } else {
    const triggers = await client.collection('io.cozy.triggers').all()
    const accountsWithoutTrigger = await fetchAccountsWithoutTriggers(
      client,
      triggers.data
    )
    const rescuableAccount = accounts.getRescuableAccount(
      accountsWithoutTrigger,
      konnector,
      userCredentials
    )
    if (rescuableAccount) {
      accountToSave = accounts.mergeAuth(rescuableAccount, userCredentials)
    } else {
      accountToSave = accounts.build(konnector, userCredentials)
    }
  }

  if (onAccountCreation) {
    logger.debug(
      `Using konnector policy "${konnectorPolicy.name}"'s custom  account creation`
    )
    accountToSave = await onAccountCreation({
      client,
      flow,
      account: accountToSave,
      konnector
    })
  }

  if (cipher && saveInVault) {
    accountToSave = accounts.setVaultCipherRelationship(
      accountToSave,
      cipher.id
    )
  } else {
    logger.warn(
      'No cipher passed when creating/updating account, account will not be linked to cipher'
    )
  }
  return await saveAccount(client, konnector, accountToSave)
}

/**
 * Event hub to launch and follow a konnector job.
 *
 * - Creates the job for a trigger
 * - Listens to changes on
 *   - the created job
 *   - the account linked to the trigger
 *
 * Transforms low-level events happening on the documents to higher-level events
 * closer to business logic.
 *
 * This should be the go to source of truth for the state of a Konnector Job.
 */
export class ConnectionFlow {
  constructor(client, trigger, konnector) {
    this.client = client
    this.trigger = trigger
    this.konnector = konnector
    this.unsubscribeAllRealtime = null

    // Bind methods used as callbacks
    this.getTwoFACodeProvider = this.getTwoFACodeProvider.bind(this)
    this.getKonnectorSlug = this.getKonnectorSlug.bind(this)
    this.handleAccountUpdated = this.handleAccountUpdated.bind(this)
    this.handleJobUpdated = this.handleJobUpdated.bind(this)
    this.handleTwoFA = this.handleTwoFA.bind(this)
    this.launch = this.launch.bind(this)
    this.sendTwoFACode = this.sendTwoFACode.bind(this)
    this.unwatch = this.unwatch.bind(this)

    this.state = {
      status: IDLE
    }

    // Stores data necessary for custom connection flow, for example
    // a budget-insight temporary token
    this.data = {}

    this.twoFAWaiters = this.twoFAWaiters || []

    this.realtime = new Realtime({ client })
  }

  getTwoFACodeProvider() {
    return accounts.getTwoFACodeProvider(this.account)
  }

  getKonnectorSlug() {
    return this.trigger.message.konnector
  }

  handleTwoFA(prevAccount) {
    const account = this.account

    const prevState = prevAccount && prevAccount.state
    const state = account.state

    if (prevState === state) {
      return
    }
    if (accounts.isTwoFANeeded(state)) {
      this.setState({ status: TWO_FA_REQUEST })
      this.emit(TWO_FA_REQUEST_EVENT)
    } else if (accounts.isTwoFARetry(state)) {
      this.setState({ status: TWO_FA_MISMATCH })
      this.emit(TWO_FA_MISMATCH_EVENT)
    }
  }

  triggerEvent(eventName, ...args) {
    logger.debug(`TriggerEvent ${eventName}`, args)
    if (isStepEvent(eventName)) {
      this.setState({ [eventName]: true })
    }
    if (eventToStatus[eventName]) {
      logger.debug(`Setting status ${eventToStatus[eventName]}`)
      this.setState({ status: eventToStatus[eventName] })
    }
    if (eventName === ERROR_EVENT) {
      this.refetchTrigger()
    }
    this.emit(eventName, ...args)
  }

  getAccount() {
    return this.account
  }

  async saveTwoFARequest(twoFARequestOptions) {
    logger.debug('Saving 2FA request', twoFARequestOptions)
    this.setState({ status: WAITING_TWOFA })
    try {
      const account = accounts.updateTwoFAState(
        this.account,
        twoFARequestOptions
      )
      await this.saveAccount(account)
      this.triggerEvent(TWO_FA_REQUEST_EVENT, twoFARequestOptions)
    } catch (error) {
      logger.error(error)
      this.setState({ status: ERRORED, error })
    }
  }

  /**
   * Waits until two fa code has been sent
   *
   * - Monitors the account (for the case where the 2FA code is filled by the konnector job)
   * - Subscribes a "manual" waiter for when the 2FA is done completely in-browser (budget-insight
   *   for example)
   *
   */
  waitForTwoFA() {
    logger.info('Waiting for two FA')
    return new Promise(rawResolve => {
      const accountId = this.account._id
      assert(accountId, 'Cannot wait for two fa on account without id')

      const resolve = () => {
        this.realtime.unsubscribe(
          'updated',
          ACCOUNTS_DOCTYPE,
          accountId,
          handleAccountUpdate
        )
        rawResolve()
      }

      const handleAccountUpdate = account => {
        if (account && account.twoFACode) {
          resolve()
        }
      }

      const updateFromCustomPolicy = () => {
        resolve()
      }

      this.twoFAWaiters.push(updateFromCustomPolicy)

      this.realtime.subscribe(
        'updated',
        ACCOUNTS_DOCTYPE,
        accountId,
        handleAccountUpdate
      )
    })
  }

  /**
   * Saves and updates internal account
   */
  async saveAccount(updatedAccount) {
    logger.debug('Saving account')
    this.account = await saveAccount(
      this.client,
      this.konnector,
      updatedAccount
    )
    logger.info('Saved account')
    return this.account
  }

  flushTwoFAWaiters() {
    logger.debug(`Flushing ${this.twoFAWaiters.length} two fa waiters`)
    for (const callback of this.twoFAWaiters) {
      callback()
    }
    this.twoFAWaiters = []
  }

  async sendAdditionalInformation(fields) {
    this.setState({ status: RUNNING_TWOFA })
    try {
      const konnectorPolicy = this.getKonnectorPolicy()
      await konnectorPolicy.sendAdditionalInformation(this, fields)
      this.flushTwoFAWaiters()
    } catch (error) {
      logger.error(error)
      this.setState({ status: ERRORED, error })
    }
  }

  /**
   * "Sends" 2FA Code
   */
  async sendTwoFACode(code) {
    this.setState({ status: RUNNING_TWOFA })
    try {
      logger.debug(`Sending two fa code ${code}`)
      await this.saveAccount(accounts.updateTwoFaCode(this.account, code))
    } catch (error) {
      // eslint-disable-next-line no-console
      logger.error(error)
      this.setState({ status: ERRORED, error })
    }
  }

  handleLoginSuccess() {
    this.jobWatcher.handleSuccess()
  }

  handleLoginSuccessHandled() {
    this.jobWatcher.disableSuccessTimer()
  }

  /**
   * - Creates io.cozy.accounts
   * - Links cipher to account
   * - Saves account
   * - Ensures trigger is existing for account
   * - Finds cipher via identifier / password
   * - Ensures a cipher is created if vault is unlocked
   * - Launches konnector job
   */
  async handleFormSubmit(options) {
    const { konnector, cipherId, vaultClient, userCredentials, t } = options
    try {
      let { account, trigger } = options

      const client = this.client

      this.setState({ status: CREATING_ACCOUNT })
      this.trigger = trigger
      this.account = account
      this.konnector = konnector
      this.t = t

      assert(client, 'No client')
      const konnectorPolicy = findKonnectorPolicy(konnector)
      logger.log(
        `Handling submit, with konnector policy ${konnectorPolicy.name}`
      )

      let cipher
      if (konnectorPolicy.saveInVault) {
        cipher = await createOrUpdateCipher(vaultClient, cipherId, {
          account,
          konnector,
          userCredentials
        })
      } else {
        logger.info(
          'Bypassing cipher creation because of konnector account policy'
        )
      }

      logger.debug('Creating/updating account...', account)

      account = await createOrUpdateAccount({
        account,
        cipher,
        client,
        flow: this,
        konnector,
        konnectorPolicy,
        userCredentials
      })

      this.account = account

      logger.info(`Saved account ${account._id}`)

      await this.ensureTriggerAndLaunch(client, {
        trigger,
        account,
        konnector,
        t
      })
      this.setState({ accountError: null })
    } catch (e) {
      logger.error(e)
      this.setState({ accountError: e })
      this.triggerEvent(ERROR_EVENT, e)
      throw e
    }
  }

  handleAccountUpdated(account) {
    const prevAccount = this.account

    // _type is not present in objects from realtime but we have to keep
    // it to be compatible with cozy-client's methods
    //
    this.account = {
      _type: prevAccount._type,
      ...account
    }
    const { state } = this.account
    if (accounts.isTwoFANeeded(state) || accounts.isTwoFARetry(state)) {
      this.handleTwoFA(prevAccount)
    } else if (accounts.isLoginSuccessHandled(state)) {
      this.handleLoginSuccessHandled()
    } else if (accounts.isLoginSuccess(state)) {
      this.handleLoginSuccess()
    }
  }

  async handleJobUpdated() {
    logger.debug('Handling update from job')
    await this.refetchTrigger()
  }

  async refetchTrigger() {
    if (!this.trigger) {
      return null
    }
    logger.debug(`Refetching trigger  ${this.trigger._id}`)
    const trigger = await fetchTrigger(this.client, this.trigger._id)
    logger.debug(`Refetched trigger`, trigger)
    this.trigger = trigger
    this.emit(UPDATE_EVENT)
  }

  async ensureTriggerAndLaunch(client, { trigger, account, konnector, t }) {
    logger.debug('Ensuring trigger...')
    this.t = t
    const ensuredTrigger = await ensureTrigger(client, {
      trigger,
      account,
      konnector,
      t
    })
    logger.info(`Trigger is ${ensuredTrigger._id}`)
    this.trigger = ensuredTrigger
    this.emit(UPDATE_EVENT)
    await this.launch()
  }

  /**
   * Launches the job and sets everything up to follow execution.
   */
  async launch() {
    logger.info('Launching job...')
    this.setState({ status: PENDING })

    this.account = await prepareTriggerAccount(this.client, this.trigger)
    this.realtime.subscribe(
      'updated',
      ACCOUNTS_DOCTYPE,
      this.account._id,
      this.handleAccountUpdated
    )
    logger.info(`Subscribed to ${ACCOUNTS_DOCTYPE}:${this.account._id}`)

    this.job = await launchTrigger(this.client, this.trigger)
    this.realtime.subscribe(
      'updated',
      JOBS_DOCTYPE,
      this.job._id,
      this.handleJobUpdated.bind(this)
    )
    this.jobWatcher = watchKonnectorJob(this.client, this.job)
    logger.info(`Subscribed to ${JOBS_DOCTYPE}:${this.job._id}`)

    for (const ev of JOB_EVENTS) {
      this.jobWatcher.on(ev, (...args) => this.triggerEvent(ev, ...args))
    }

    this.unsubscribeAllRealtime = () => {
      this.jobWatcher.unsubscribeAll()
      this.realtime.unsubscribeAll()
    }
  }

  unwatch() {
    if (typeof this.unsubscribeAllRealtime === 'function') {
      this.unsubscribeAllRealtime()
    }
  }

  reset() {
    this.trigger = null
    this.job = null
    this.emit(UPDATE_EVENT)
  }

  getDerivedState() {
    const trigger = this.trigger
    const { status, accountError } = this.state
    const triggerError = triggersModel.getKonnectorJobError(trigger)
    return {
      running: ![ERRORED, IDLE, SUCCESS].includes(status),
      twoFARunning: status === RUNNING_TWOFA,
      twoFARetry: status == TWO_FA_MISMATCH,
      triggerError: triggerError,
      trigger,
      accountError,
      error: accountError || triggerError,
      konnectorRunning: triggersModel.isKonnectorRunning(trigger)
    }
  }

  getState() {
    return {
      ...this.state,
      ...this.getDerivedState()
    }
  }

  setState(update) {
    this.state = { ...this.state, ...update }
    this.emit(UPDATE_EVENT)
  }

  setData(update) {
    this.data = { ...this.data, ...update }
  }

  getData() {
    return this.data
  }

  getKonnectorPolicy() {
    return findKonnectorPolicy(this.konnector)
  }

  getAdditionalInformationNeeded() {
    const konnectorPolicy = this.getKonnectorPolicy()
    if (!konnectorPolicy) {
      throw new Error(
        'A konnector policy is needed to be able to detect additional information'
      )
    } else {
      const fields = konnectorPolicy.getAdditionalInformationNeeded(this)
      return fields
    }
  }
}

MicroEE.mixin(ConnectionFlow)

export default ConnectionFlow
