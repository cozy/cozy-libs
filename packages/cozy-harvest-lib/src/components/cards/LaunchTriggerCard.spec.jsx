import LaunchTriggerCard from 'components/cards/LaunchTriggerCard'
import { mount } from 'enzyme'
import React from 'react'

import CozyClient, { CozyProvider } from 'cozy-client'
import I18n from 'cozy-ui/transpiled/react/providers/I18n'

import enLocale from '../../locales/en.json'
import ConnectionFlow from '../../models/ConnectionFlow'

jest.mock('../../models/ConnectionFlow', () => {
  // Require the original module to not be mocked...
  const { default: mockConnectionFlow } = jest.requireActual(
    '../../models/ConnectionFlow'
  )

  mockConnectionFlow.prototype.watchJob = jest.fn()

  return mockConnectionFlow
})

const triggerFixture = {
  _id: 'd861818b62204988bf0bb78c182a9149',
  arguments: '0 0 0 * * 0'
}
const konnectorFixture = {
  slug: 'boursorama83',
  parameters: {
    bankId: '100000'
  },
  partnership: {
    domain: 'https://budget-insight.com'
  }
}

describe('LaunchTriggerCard', () => {
  const client = new CozyClient({})
  client.plugins = {
    realtime: {
      sendNotification: jest.fn(),
      subscribe: jest.fn()
    }
  }

  const setup = ({ props }) => {
    const root = mount(
      <CozyProvider client={client}>
        <I18n dictRequire={() => enLocale} lang="en">
          <LaunchTriggerCard {...props} />
        </I18n>
      </CozyProvider>
    )
    return { root }
  }

  it('should render impossible to run message when konnector is clientSide without an accessible launcher', () => {
    const { root } = setup({
      props: {
        flow: new ConnectionFlow(
          client,
          triggerFixture,
          {
            slug: 'test',
            name: 'testname',
            clientSide: true
          },
          konnectorFixture
        )
      }
    })
    expect(root.html()).toMatchSnapshot()
  })
})
