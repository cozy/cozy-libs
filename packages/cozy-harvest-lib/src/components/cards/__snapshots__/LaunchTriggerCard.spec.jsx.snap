// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchTriggerCard should render impossible to run message when konnector is clientSide without an accessible launcher 1`] = `"<div class=\\"styles__c-card___YgP7B\\"><div class=\\"u-flex u-flex-column-s\\"><ul class=\\"u-nolist u-m-0 u-mr-1 u-pl-0 u-flex-grow-1\\"><li class=\\"u-mb-half\\"><span class=\\"MuiTypography-root u-mr-half MuiTypography-caption MuiTypography-colorTextSecondary\\" style=\\"display: inline-block;\\">Update:</span><span class=\\"MuiTypography-root MuiTypography-caption MuiTypography-colorTextPrimary\\" style=\\"display: inline-block;\\">Last sync. unknown</span></li><div class=\\"styles__media___1rIBu styles__media--top___1t0j5\\"><div class=\\"styles__img___3sztD u-m-1\\"><svg viewBox=\\"0 0 16 16\\" class=\\"styles__icon___23x3R\\" width=\\"16\\" height=\\"16\\"><defs><path id=\\"info_svg__a\\" d=\\"M8 16A8 8 0 118 0a8 8 0 010 16zM7 4a1 1 0 102 0 1 1 0 10-2 0zm1 2H6v2h1v4a1 1 0 001 1h2v-2H9V7a1 1 0 00-1-1z\\"></path></defs><use fill-rule=\\"evenodd\\" xlink:href=\\"#info_svg__a\\"></use></svg></div><div class=\\"styles__bd___3SAX2 u-m-1\\"><p class=\\"MuiTypography-root MuiTypography-body1 MuiTypography-colorTextPrimary\\">To synchronize your testname data in your Cozy you must use the Cozy app on your phone or tablet.</p></div></div></ul></div></div>"`;

exports[`LaunchTriggerCard should render normally when konnector is not clientSide 1`] = `"<div class=\\"styles__c-card___YgP7B\\"><div class=\\"u-flex u-flex-column-s\\"><ul class=\\"u-nolist u-m-0 u-mr-1 u-pl-0 u-flex-grow-1\\"><li class=\\"u-mb-half\\"><span class=\\"MuiTypography-root u-mr-half MuiTypography-caption MuiTypography-colorTextSecondary\\" style=\\"display: inline-block;\\">Update:</span><span class=\\"MuiTypography-root MuiTypography-caption MuiTypography-colorTextPrimary\\" style=\\"display: inline-block;\\">Last sync. unknown</span></li><li class=\\"u-mb-half\\"><span class=\\"MuiTypography-root u-mr-half MuiTypography-caption MuiTypography-colorTextSecondary\\" style=\\"display: inline-block;\\">Frequency:</span><span class=\\"MuiTypography-root MuiTypography-caption MuiTypography-colorTextPrimary\\" style=\\"display: inline-block;\\">Once a week</span></li></ul><div><button style=\\"line-height: 1.4;\\" type=\\"submit\\" class=\\"styles__c-btn___3kXsk styles__c-btn--subtle___2rRQ0 styles__c-btn--center___Nny0n u-mh-0 u-mv-0\\"><span><svg viewBox=\\"0 0 16 16\\" class=\\"styles__icon___23x3R\\" width=\\"16\\" height=\\"16\\" focusable=\\"false\\"><path fill-rule=\\"evenodd\\" d=\\"M16 7.2V3c0-.6-.4-1-1-1-.5 0-.8.3-.9.7A8.09 8.09 0 008 0C4.9 0 2 1.8.7 4.6c-.2.5 0 1.1.5 1.3.5.2 1.1 0 1.3-.5C3.5 3.4 5.6 2 8 2c1.5 0 3 .6 4 1.5-.3 0-.6.1-.8.3-.4.4-.3 1.1.1 1.4l3 2.5h.1c.1.1.2.1.3.1h.4c.1 0 .2-.1.3-.1h.1c.1-.1.2-.1.2-.2.2 0 .2-.1.3-.3 0 .1 0 0 0 0m-1.2 2.9c-.5-.2-1.1 0-1.3.5A6 6 0 018 14c-1.5 0-2.9-.6-4-1.5.3 0 .5-.1.7-.4.4-.4.3-1.1-.1-1.4l-3-2.5h-.1L1.3 8h-.1c-.2 0-.5 0-.7.1H.4c-.1.1-.1.2-.2.3-.1.1-.1.2-.2.3V13c0 .6.4 1 1 1 .5 0 .9-.3 1-.8C3.5 14.9 5.6 16 8 16c3.1 0 5.9-1.8 7.3-4.5.2-.5 0-1.1-.5-1.4\\"></path></svg><span>Synchronize</span></span></button></div></div></div>"`;
